posts/2021-05-07-caught-red-handed.md

```diff
@@ -0,0 +1,230 @@
+
+Another recent example looks [like this](https://github.com/rust-lang/rust/issues/85039):
+
+```
+thread 'rustc' panicked at 'found unstable fingerprints for predicates_of(<massive text describing rustc internals elided>)', /rustc/.../compiler/rustc_query_system/src/query/plumbing.rs:593:5
+```
+
+They all arise from inconsistencies when comparing the incremental-compilation cache stored on disk against the values computed during a current `rustc` invocation, which means they all arise from using incremental compilation.
+
+There are several ways that you may have incremental compilation turned on:
+
+1. You may be building with the `dev` or `test` [profiles][] which default to having incremental compilation enabled.
+2. You may have set the [environment variable][env-vars] `CARGO_INCREMENTAL=1`
+3. You may have enabled the `build.incremental` [setting in your Cargo config][cargo-config]
+4. You may have enabled the `incremental` [setting in your Cargo.toml][cargo-toml] for a given profile
+
+[env-vars]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
+[cargo-config]: https://doc.rust-lang.org/cargo/reference/config.html#buildincremental
+[cargo-toml]: https://doc.rust-lang.org/cargo/reference/profiles.html#incremental
+[profiles]: https://doc.rust-lang.org/cargo/reference/profiles.html
+
+If your project has not adjusted the defaults, then when running `cargo build
+--release` or otherwise in the `release` profile configuration incremental is
+disabled on all Rust versions[^1], and these issues should not affect your release
+builds.
+
+## What should a Rust programmer do in response
+
+The Internal Compiler Error asks you to report a bug, and if you can do so, we still want that information. We *want* to know about the cases that are failing.
+
+But regardless of whether or not you file a bug, the problem here can be resolved by either:
+
+ 1. upgrading to 1.52.1, if you have not yet done so (which will disable
+    incremental for you).
+ 2. deleting your incremental compilation cache (e.g. by running `cargo clean`), or
+ 3. force incremental compilation to be disabled, by setting `CARGO_INCREMENTAL=0` in your environment or `build.incremental` to `false` in the `config.toml`.
+
+We recommend that users of 1.52.0 upgrade to 1.52.1, which disables incremental
+compilation.
+
+We do *not* recommend that users of 1.52.0 downgrade to an earlier version of Rust in response to this problem. As noted above, there is at least one instance of a silent [miscompilation][issue-82920] caused by incremental compilation that was not caught until we added the fingerprint checking.
+
+If a user is willing to deal with the incremental verification ICE's, and wishes
+to opt back into the 1.52.0 behavior, they may set `RUSTC_FORCE_INCREMENTAL` to
```

Comment:
For these users who are willing to put up with some ICEs, would it make sense to ask them to install the 1.53 beta? We have fixed a number in beta although not all of them.

Reply:
(The text a few paragraphs down does suggest that option, of trying out the 1.53 beta since a number of bugs are fixed there.)